packages/flutter_tools/lib/src/compile.dart

```diff
@@ -98,14 +98,36 @@ class StdoutHandler {
 
   bool _suppressCompilerMessages;
   bool _expectSources;
+  bool _badState = false;
 
   void handler(String message) {
     printTrace('-> $message');
+    if (_badState) {
+      return;
+    }
     const String kResultPrefix = 'result ';
     if (boundaryKey == null && message.startsWith(kResultPrefix)) {
       boundaryKey = message.substring(kResultPrefix.length);
       return;
     }
+    // Invalid state, see commented issue below for more information.
+    if (boundaryKey == null) {
+      // This needs to complete error so that the exceptions are not swallowed
```

Comment:
As discussed offline, update the comment to mention that throwing a synchronous exception ('throw ToolExit()') would fail to cancel the stream, leading to multiple errors being reported until the ToolExit exception is handled, reported, and the program really exits.